### ISSUE TITLE
Quick fix introduces bugs by renaming with ignoring name conflicts.

### ISSUE THREAD

**Issue by polina-c** [dart-lang/sdk#52157]
Lint: https://dart-lang.github.io/linter/lints/no_leading_underscores_for_local_identifiers.html

https://user-images.githubusercontent.com/12115586/234104118-e2b10732-a2fb-4389-9679-e6bd2a3ea4bd.mov

**Comment by srawlins** [comment]
Thanks for filing an issue, @polina-c, can you provide a snippet of reproduction code (hard for me to hand-type code from a video), and explain each of the bugs which are introduced when you execute the quick fix?

**Comment by scheglov** [comment]
FWIW, I was going to make the same comment - it is hard to infer what is the problem from video.

Here renamed `_foo` will shadow `A.foo`.
```dart
class A {
  int foo = 0;

  void bar(int _foo) {
    foo + _foo;
  }
}
```

Here renamed `_foo` will be shadowed by local `foo`.
```dart
void bar(int _foo) {
  var foo = 0;
  foo + _foo;
}
```

**Comment by bwilkerson** [comment]
Are we proposing to always disallow the refactoring when it would introduce a shadowing, or are we proposing to sometimes add `this.` to preserve the previous behavior?

**Comment by scheglov** [comment]
When inside a setter in a class, I think it is usual to use `this.foo = foo;`.
This does not work though for top-level properties.

I think adding `this.` while removing leading `_` would be too magical.
We might consider adding a separate quick assist to prepend `this.` to all references to a field.

**Comment by polina-c** [comment]
Here is the code: https://github.com/flutter/devtools/blob/617377e579b9525d528fe354a183c5221e5fd10c/packages/devtools_app/lib/src/shared/analytics/_analytics_web.dart#L618

**Comment by polina-c** [comment]
Actually, when i do the same renaming with F2, VSCode shows me the waring about the name conflict. It is strange that the lint misses it. 
I thought lints use the same API for renaming as F2. Do they?

**Comment by bwilkerson** [comment]
F2 is probably invoking the rename refactoring, which will potentially look across all of the files in your package to update the names. The quick fix only applies to local variables, so we know we only need to look in the current scope. The second is a much simpler problem than the first, so we have a different implementation for it. So no, they don't use the same API.

**Comment by srawlins** [comment]
Drive-by comment: @FMorschel fixed a bunch of these a few months back; not sure about this one.

**Comment by FMorschel** [comment]
I fixed some renaming things, not related to this quick fix. 

> I think adding `this.` while removing leading `_` would be too magical.

Not even our renames would fix this case. They only would if we were renaming the class member (where it could add the `this.` up front) and it would (otherwise) conflict with the local variable.

**Comment by FMorschel** [comment]
After running into https://github.com/dart-lang/sdk/issues/63862 today, I thought a bit more about this. What I think the best approach we _currently_ can take would be to have a more specific diagnostic code for shadowing cases like the above, and not to display the rename quick-fix for those.

I'll try to look at this one.

**Comment by FMorschel** [comment]
Is there any way we could tell the user where the other variable is? Maybe not with the quick-fix, but maybe the rename, when warned, could have a button to somehow tell the user where the other variable is. If that is a big `build` method, these can be hard to spot.

CC @DanTup for ideas too

**Comment by bwilkerson** [comment]
I'd love to be wrong, but as far as I'm aware, the best we could do in this case is to include that information in the message we send back when asking whether the user wants to continue. And the best we could do there is to point to it using something like a line number. Something like "Renaming '_foo' to 'foo' would conflict with the definition of 'foo' on line 623."

**Comment by FMorschel** [comment]
I wish we could have something like context messages for this too 😅

**Comment by DanTup** [comment]
I think Brian is correct - I don't think there's any good way to provide a location here (at least not in a standard LSP way). For renames (F2), we do already support prompting in some cases and asking if you want "Rename anyway?". We don't support this for quick-fixes because we send the edit to the client in the codeAction request so there's no opportunity to prompt the user.

However, we do support this for refactors because we don't ship the edits inline for them, we compute them later via a command when invoked. In theory, if we made quick-fixes also work that way (which the kinda do if the client doesn't support codeActionLiterals), then we could potentially add prompts in that case too.

(although, in the case of the quick-fix I wonder whether it might be better to pick a unique name if the usual target is taken, rather than asking the user if we want to break the code?)

**Comment by FMorschel** [comment]
Maybe if we detect the conflict, we could instead of simply removing `_` append a `Local` or something? So the diagnostic goes away but the user can rename to something else then after the fact? And if that name is taken append numbers? I know that's probably not what the user would do, but it may help removing the diagnostic 🤷‍♂️

**Comment by bwilkerson** [comment]
Yeah, I wouldn't recommend or support doing that. The goal isn't to resolve the diagnostic, the goal is to make it easier for the user to correct the problem (or, more formally, to provide tools that can increase the user's velocity). But whether the user has to rename `_foo` or `local_foo`, the amount of work is the same. So renaming it to some other likely-to-be-wrong name isn't really helping.

**Comment by DanTup** [comment]
> But whether the user has to rename `_foo` or `local_foo`, the amount of work is the same.

Perhaps I've misunderstood, but at 9s in to the video at the top on line 620, hasn't the rename resulted in broken code (`_foo = _foo` where both identifiers were different elements before)? If so, renaming again will presumably (incorrectly) update both of them and still leave incorrect code? Whereas if the assigned name had been unique, the code would still be correct, and a manual rename would only rename the expected names?

**Comment by FMorschel** [comment]
Yes, exactly, but if we simply didn't rename anything when we detected a case like this, the user would need to remame manually and we wouldn't edit anything to what the user didn't mean to write (adding `local`). And the problem with the missing `this.` would still happen. 

I think it's best we do that and create other refactor tools to help the user with this process instead of guessing what they meant to name their variables.

**Comment by FMorschel** [comment]
For anyone interested, I'm working on:

- https://github.com/dart-lang/sdk/issues/63833

The CL is up, and it's big, but hopefully it also helps with this, somehow.

---

Also, I've opened:

- https://github.com/dart-lang/sdk/issues/63909

**Comment by FMorschel** [comment]
Okay. I've done some work around the lint now. Seems to be working as I expected. I do have a clarifying question on the quick fix, then. After that, I'll upload the change. I wasn't aware of this behaviour.

I saw this test case:

```dart
void f() {
  var _foo = 1;
  var foo = true;
  print(_foo);
  print(foo);
}
```

For this, we would compute:

```dart
void f() {
  var foo0 = 1; //suffixed 0
  var foo = true;
  print(foo0); //suffixed 0
  print(foo);
}
```

Should I keep the same behaviour but do that more consistently now, @bwilkerson? Or should I simply avoid renaming and let the user handle it manually?

Edit:

My opinion is the same as above. We should let the user edit these cases manually and only fix the ones we know would be fine to handle.

**Comment by bwilkerson** [comment]
If you're saying what I said at https://github.com/dart-lang/sdk/issues/52157#issuecomment-5047504627, then I agree. The fix makes sense when removing the underscore is a valid way to resolve the issue because the remaining name is likely to still be semantically valid. Any other change can be done using 'rename' and we shouldn't offer a quick fix.

**Comment by FMorschel** [comment]
I forgot to post it back here. Here is the CL for this: https://dart-review.googlesource.com/c/sdk/+/530700
